### PR TITLE
mrc-3603: Migrate serializedprofile column to JSON serialization

### PR DESCRIPTION
**Pull request by r-ash**
I'm not expecting this to be merged this is a 1 off script and I am just putting it here as a reference. I can move this to somewhere else if there is somewhere more appropriate.

We recently updated from pac4j 3.3.0 to 5.4.3 see https://github.com/mrc-ide/hint/pull/731/files

This update includes some changes to format for the serialised user profile in the database see https://github.com/pac4j/pac4j/blob/master/documentation/blog/what_s_new_in_pac4j_v5.md#b-serializedprofile

In particular 3.3.0 was using Java serialization and now pac4j uses JSON. This was resulting in errors on our end when trying to login as test.user on staging and preview (see ticket for errors). We wiped the volumes on dev deployment so the test.user there has serializedprofile in the new format. We can't just wipe volume on prod so we need to migrate to the new format as this will affect all existing users.

The uses https://github.com/pac4j/pac4j/blob/a1ae387ce42b09ffa114065f8782061404be4cf4/pac4j-core/src/main/java/org/pac4j/core/util/serializer/ProfileServiceSerializer.java to decode serialized profiles and recode them in JSON format.

Process for running on prod
1. Make a backup `./backup/backup`
2. Checkout this branch of hint-deploy
3. Run the migration `cd scripts & ./scripts/run_fix_profile_encoding.sh`
4. Check login now works and switch back to master branch of hint-deploy

I have run this on staging, if you want to you can restore staging from the backup and re-run this to see it working.

One question:
In the old version of pac4j `canAttributesBeMerged` was FALSE see https://github.com/pac4j/pac4j/blob/pac4j-3.9.0/pac4j-core/src/main/java/org/pac4j/core/profile/UserProfile.java#L48
But in new pac4j `canAttributesBeMerged` is TRUE see https://github.com/pac4j/pac4j/blob/master/pac4j-core/src/main/java/org/pac4j/core/profile/BasicUserProfile.java#L48

So after the migration, newly added accounts have `canAttributesBeMerged` true, whilst the migrated ones have `canAttributesBeMerged` false. Is this ok? Should I update the old ones during the migration to set this to true?

### CODE REVIEW

**Comment by hillalex** [comment]
So on prod you'll just manually pull the db password out of the vault and change the hard-coded value?

I don't know exaclty how the `canAttributesBeMerged` property of profiles is used, but in HINT afaik the profiles are very minimal things and are never mutated, so my guess is this property is never invoked.

**Comment by LekanAnni** [comment]
> So after the migration, newly added accounts have `canAttributesBeMerged` true, whilst the migrated ones have `canAttributesBeMerged` false. Is this ok? Should I update the old ones during the migration to set this to true?

according to pac4j https://www.pac4j.org/apidocs/pac4j/3.3.0/org/pac4j/core/profile/UserProfile.html protected UserProfile(boolean canAttributesBeMerged)
Create a profile with possibility to merge attributes with the same name and collection-type values.

Parameters:
canAttributesBeMerged - if true - merge attributes with the same name and collection-type values, if false - overwrite them

I believe this should be set to true.

**Comment by LekanAnni** [comment]
> So on prod you'll just manually pull the db password out of the vault and change the hard-coded value?

Not password but profile attributes, we don’t directly store profile attributes but pac4j automatically serialises user profile into User table when an account is created via DBProfileService.

**Review by LekanAnni — APPROVED** [review]
LGTM 👌🏾 

You should run update in a try catch and finally to close database connection. 

I think you should add detailed comment, explaining what the code does.  Could also add script to util, we may need it again in future or for other projects.

File: scripts/fix_profile_encoding.groovy [hunk 1]
```diff
@@ -0,0 +1,55 @@
+@GrabConfig(systemClassLoader=true)
+@Grab(group='org.postgresql', module='postgresql',  version='9.4-1205-jdbc42')
+@Grab(group='org.pac4j', module='pac4j-core', version='3.9.0')
```
[the remark this answers is not in the document]
Comment (r-ash) [reply scripts/fix_profile_encoding.groovy:3]: Yeah this is where I had the most pain. Only 3.9 has the right serialVersionUID to decode the profiles and has the ProfileServiceSerializer available

[the remark this answers is not in the document]
Comment (hillalex) [reply scripts/fix_profile_encoding.groovy:3]: Well done Rob, this looks like it was a proper pain to track down!

**Comment by r-ash** [comment]
> Parameters: canAttributesBeMerged - if true - merge attributes with the same name and collection-type values, if false - overwrite them
> 
> I believe this should be set to true.

Ok so on review this is actually a private final with no accessor https://github.com/pac4j/pac4j/blob/73ecc89dc2c40c37385184d8db9a16790f30bea1/pac4j-core/src/main/java/org/pac4j/core/profile/UserProfile.java#L45 so I think I cannot update it for old profiles. Sounds like it shouldn't cause problems if it is set to false

**Review by Nicolas-Dolan — COMMENTED** [review]
Not really sure what I'm looking at, but it's groovy

File: scripts/fix_profile_encoding.groovy [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        this.serializedProfile = serializer.encode(decoded)
+        println "Migrated to JSON format"
+        println this.serializedProfile
+        this.updated = true;
+    }
+}
+
+def dbUrl      = "jdbc:postgresql://hint_db/hint"
+def dbUser     = "hintuser"
+def dbPassword = "changeme"
```
[the remark this answers is not in the document]
Comment (r-ash) [reply scripts/fix_profile_encoding.groovy:48]: Should be ideally yeah, but this is already in plaintext in this repo when the db is started so I think this is not adding much additional risk

**Pull request merged**